File: 3ero/concurrente/practica_2/p4/p4.rb
require 'concurrent'

ALUMNOS = (ARGV[0] || 40).to_i

tareas = ALUMNOS.times.collect { Concurrent::Semaphore.new(1) }
mutex = ALUMNOS.times.collect { Concurrent::Semaphore.new(0) }
correcciones = Array.new(ALUMNOS, false)
entregas = []
cant = ALUMNOS

ALUMNOS.times do | i |
  Thread.new do
    while !correcciones[i]
      tareas[i].acquire # Funcion P del 'lenguaje' de concurrente
      sleep 1 # Se hace tarea
      puts "entregar tarea alumno #{i} "
      entregas.push i
      tareas[i].release # Funcion V del 'lenguaje' de concurrente
      mutex[i].acquire
    end
    puts "alumno #{i}: chau!"
  end
end


maestra = Thread.new do
  while cant != 0
    if !entregas.empty?
      entregas.pop.tap do |id|
        tareas[id].acquire
        [true, false].sample.tap do |correccion| # Se corrige la tarea al azar
          correcciones[id] = correccion
          cant -= 1 if correccion
          puts "correccion tarea #{id} = #{correccion} - faltan #{cant} "
          tareas[id].release
          mutex[id].release
        end
      end
    end
  end
end

maestra.join # Se deja el hilo como el principal

File: 3ero/2do_Semestre/concurrente/Practicas/practica_2/p4/concurrente_like/p4.c
sem tarea[40] = ([40], 0);
sem mutex[40] = ([40], 0);
boolean correcciones[40] = ([40], false);
cola entregas;
int cant = 40;

Process alumnos[i = 1 to 40]
{
  while(!correcciones[i]){
    P(tareas[i]);
    // Hago tarea
    encolar(entregas, i);
    V(tareas[i]);
    P(mutex[id]);
  }
}

Process maestra
{
  for(int i = 1; i < 40; i++){
    V(tarea[i]);
  }
  while(cant != 0){
    if(!entregas.empty?) {
      int id = desencolar(entregas);
      P(tareas[id]);
      if(//tarea es correcta){
        cant -= 1;
        correcciones[id] = true;
      }
      V(tarea[id]);
      V(maestra[id]);
    }
  }
}
